python/pyxet/pyxet/file_operations.py

```diff
@@ -0,0 +1,370 @@
+        if _isdir(dest_fs, dest_dir):
+          dest_dir = None # No need to create this
+
+        cp_unit = CopyUnit(src_path=src_path, dest_path=dest_path, dest_dir = dest_dir, size = size_hint)
+                           
+    _single_file_copy_impl(cp_unit, src_fs, dest_fs)
+
+
+def _build_cp_action_list_impl(src_fs, src_path, dest_fs, dest_path, recursive):
+    
+    
+    
+    # This function has two parts.  First, we we set a number of variables that then determine how 
+    # the second section will behave.  
+
+    dest_is_xet = dest_fs.protocol == "xet"
+
+    # If both the source and the end is a xet, then we can copy directories as entire units. 
+    dirs_directly_copyable = dest_is_xet and src_fs.protocol == "xet"
+
+    # If the destination is specified as a directory, then we want to respect that, putting
+    # things as needed within that directory or erring out if it exists but is not itself a 
+    # directory.
+    dest_specified_as_directory = dest_path.endswith("/") 
+    dest_path = dest_path.rstrip("/")
+
+
+    # Now set the following variables depending on src_path and dest_path, and what dest_path is
+    # on the remote.
+    file_filter_pattern = None  # If not none, only files matching this pattern will be copied.
+    src_info = None  # Info about the source, if applicable.  Saves calls to info, which may be expensive. 
+    src_is_directory = None     # True if the src is a directory.  This must be set after wildcards are detected.
+    dest_dir = None   # The destination directory or branch containing the dest_path. 
+
+    if '*' in src_path:
+        # Handling wildcard cases.  
+        
+        # The file card matching works by recasting the operation as a directory -> directory copy, 
+        # but with a possible filter pattern. 
+
+        # validate
+        # we only accept globs of the for blah/blah/blah/[glob]
+        # i.e. the glob is only in the last component
+        # src_root_dir should be blah/blah/blah here
+        src_root_dir, end_component = _path_split(src_fs, src_path)
+
+        if '*' in src_root_dir:
+            raise ValueError(
+                f"Invalid glob {src_path}. Wildcards can only appear in the last position")
+
+        src_path = src_root_dir 
+        src_is_directory = True # Copying the contents of the new source directory 
+        file_filter_pattern = end_component if end_component != "*" else None
+
+        dest_specified_as_directory = True
+
+        dest_dir=dest_path  # Containing dir.
+        dest_path=None  # In this case, the path information will be taken from the source file name.
+
+    else:
+
+        # Validate that the source path is specified correctly, and set src_is_directory. 
+        if src_path.endswith("/"):
+            src_path = src_path.rstrip("/")
+            src_info = src_fs.info(src_path)
+            src_is_directory = True
+            if not src_info['type'] in ["directory", "branch"]: 
```

Comment:
Let's update all these manual checks to call `_isdir`

Reply:
So the issue is that _isdir performs a call to fs.info, which is a remote call to the server if it's a remote file system.  So ideally we should limit calls to that if we can get it another way.  Here we need the size, so making one call to fs.info for that, or obtaining it from the find method that returns it in bulk, would be way faster.

Reply:
Good point